dot/sync/chain_sync.go

```diff
@@ -53,693 +48,630 @@ func (s chainSyncState) String() string {
-		}
+	if parentExists {
+		request = singleBlockRequest(announcedHash, bootstrapRequestData)
+	} else {
+		gapLength = uint32(announcedHeader.Number - highestFinalizedHeader.Number)
+		startAtBlock = highestFinalizedHeader.Number + 1
+		request = descendingBlockRequest(announcedHash, gapLength, bootstrapRequestData)
 	}
-}
 
-func (cs *chainSync) ignorePeer(who peer.ID) {
-	if err := who.Validate(); err != nil {
-		return
-	}
+	logger.Debugf("received a block announce from %s, requesting %d blocks, starting %s (#%d)",
+		peerWhoAnnounced, gapLength, announcedHash, announcedHeader.Number)
 
-	cs.Lock()
-	cs.ignorePeers[who] = struct{}{}
-	cs.Unlock()
+	resultsQueue := make(chan *syncTaskResult)
+	wg := sync.WaitGroup{}
+
+	wg.Add(1)
+	go cs.handleWorkersResults(resultsQueue, startAtBlock, gapLength, &wg)
+	cs.workerPool.submitBoundedRequest(request, peerWhoAnnounced, resultsQueue)
+	wg.Wait()
+
+	return nil
 }
 
-func (cs *chainSync) sync() {
-	// set to slot time
-	ticker := time.NewTicker(cs.slotDuration)
+func (cs *chainSync) requestPendingBlocks() error {
+	logger.Infof("total of pending blocks: %d", cs.pendingBlocks.size())
+	if cs.pendingBlocks.size() == 0 {
+		return nil
+	}
 
-	for {
-		select {
-		case ps := <-cs.workQueue:
-			cs.maybeSwitchMode()
+	highestFinalized, err := cs.blockState.GetHighestFinalisedHeader()
+	if err != nil {
+		return fmt.Errorf("getting highest finalised header: %w", err)
+	}
 
-			if err := cs.handleWork(ps); err != nil {
-				logger.Errorf("failed to handle chain sync work: %s", err)
-			}
-		case res := <-cs.resultQueue:
-			if err := cs.handleResult(res); err != nil {
-				logger.Errorf("failed to handle chain sync result: %s", err)
-			}
-		case <-ticker.C:
-			cs.maybeSwitchMode()
+	pendingBlocks := cs.pendingBlocks.getBlocks()
+	for _, pendingBlock := range pendingBlocks {
+		if pendingBlock.number <= highestFinalized.Number {
+			cs.pendingBlocks.removeBlock(pendingBlock.hash)
+			continue
+		}
+
+		parentExists, err := cs.blockState.HasHeader(pendingBlock.header.ParentHash)
+		if err != nil {
+			return fmt.Errorf("getting pending block parent header: %w", err)
+		}
 
-			workers, err := cs.handler.handleTick()
+		if parentExists {
+			err := cs.handleReadyBlock(pendingBlock.toBlockData())
 			if err != nil {
-				logger.Errorf("failed to handle tick: %s", err)
-				continue
+				return fmt.Errorf("handling ready block: %w", err)
 			}
+			continue
+		}
 
-			for _, worker := range workers {
-				cs.tryDispatchWorker(worker)
-			}
-		case fin := <-cs.finalisedCh:
-			// on finalised block, call pendingBlocks.removeLowerBlocks() to remove blocks on
-			// invalid forks from the pending blocks set
-			cs.pendingBlocks.removeLowerBlocks(fin.Header.Number)
-		case <-cs.ctx.Done():
-			return
+		gapLength := pendingBlock.number - highestFinalized.Number
+		if gapLength > 128 {
+			logger.Criticalf("GAP LENGHT: %d, GREATER THAN 128 block", gapLength)
+			gapLength = 128
```

Comment:
Why is this critical and if it is why do we just auto reset to 128?

Reply:
This still under development since it is part of the tip sync implementation, I will resolve it asap